syncer/parallel_sync.go

```diff
@@ -0,0 +1,201 @@
+package syncer
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/gateway"
+	"go.sia.tech/core/types"
+	"go.uber.org/zap"
+)
+
+func (s *Syncer) parallelSync(ctx context.Context, cs consensus.State, headers []types.BlockHeader) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	start := time.Now()
+
+	type Req struct {
+		base      types.ChainIndex
+		tip       types.ChainIndex
+		numBlocks uint64
+	}
+
+	type Resp struct {
+		blocks []types.Block
+		states []consensus.State
+		req    Req
+		err    error
+	}
+
+	const blocksPerReq = 100 // peers should never send fewer than this if available
+	reqs := make([]Req, (len(headers)+blocksPerReq-1)/blocksPerReq)
+	for i := range reqs {
+		off := uint64(i) * blocksPerReq
+		numBlocks := min(blocksPerReq, uint64(len(headers[off:])))
+		base := types.ChainIndex{
+			ID:     headers[off].ParentID,
+			Height: cs.Index.Height + off,
+		}
+		tip := types.ChainIndex{
+			ID:     headers[off+numBlocks-1].ID(),
+			Height: base.Height + numBlocks - 1,
+		}
+		reqs[i] = Req{base, tip, numBlocks}
+	}
+
+	workFn := func(p *Peer, req Req) (resp Resp) {
+		resp.req = req
+		startTime := time.Now()
+		if req.base.Height >= cs.Network.HardforkV2.RequireHeight {
+			cs, b, err := p.SendCheckpoint(req.base, cs.Network, s.config.SendBlockTimeout)
+			if err != nil {
+				return Resp{req: req, err: err}
+			}
+			cs, _ = consensus.ApplyBlock(cs, b, consensus.V1BlockSupplement{}, time.Time{})
+			blocks, _, err := p.SendV2Blocks(ctx, []types.BlockID{cs.Index.ID}, req.numBlocks, s.config.SendBlocksTimeout)
+			if err != nil {
+				return Resp{req: req, err: err}
+			} else if uint64(len(blocks)) != req.numBlocks {
+				return Resp{req: req, err: errors.New("peer returned wrong number of blocks")}
+			} else if blocks[len(blocks)-1].ID() != req.tip.ID {
+				return Resp{req: req, err: errors.New("peer returned wrong blocks")}
+			}
+			resp.blocks = blocks
+			for _, b := range blocks {
+				if err := consensus.ValidateBlock(cs, b, consensus.V1BlockSupplement{}); err != nil {
+					return Resp{req: req, err: err}
+				}
+				cs, _ = consensus.ApplyBlock(cs, b, consensus.V1BlockSupplement{}, time.Time{})
+				resp.states = append(resp.states, cs)
+			}
+		} else {
+			blocks, _, err := p.SendV2Blocks(ctx, []types.BlockID{req.base.ID}, req.numBlocks, s.config.SendBlocksTimeout)
```

Comment:
Not sure I follow the logic here. What about v1 blocks? At this point it seems like the height could be before the allow height?

Reply:
yeah, not the best name. In this case `V2` refers to the v2 *encoding* of blocks. The blocks themselves can be v1 blocks.

Now that the old `SendBlocks` is deprecated, we could rename the method to just `SendBlocks`.